src/Aspire.Hosting/DistributedApplicationResourceBuilder.cs

```diff
@@ -24,7 +24,7 @@ internal sealed class DistributedApplicationResourceBuilder<T>(IDistributedAppli
 
         // If the behavior is AddReplace then there should never be more than one annotation present. The following call will result in an exception which
         // allows us to easily spot these bugs.
-        if (behavior == ResourceAnnotationMutationBehavior.Replace && Resource.Annotations.OfType<TAnnotation>().SingleOrDefault() is { } existingAnnotation)
+        if (behavior == ResourceAnnotationMutationBehavior.Replace && Resource.Annotations.SingleOrDefault(a => a.GetType() == annotation.GetType()) is { } existingAnnotation)
```

Comment:
This behavior is different.

Reply:
Yeah .. will fix this.

Reply:
Tested added (found a bug actually). Arguably this API could be tightened up even further by making it that you can only add annotations that are sealed types, to avoid the ambiguity that inheritance introduces here.